File: .NetProject/src/JobDispatch.Api/Services/JobService.cs
using JobDispatch.Api.Models;

namespace JobDispatch.Api.Services;

public sealed class JobService : IJobService
{
    private readonly IJobRepository _repository;
    private readonly JobDispatchOptions _options;
    private readonly ILogger<JobService> _logger;

    public JobService(IJobRepository repository, JobDispatchOptions options, ILogger<JobService> logger)
    {
        _repository = repository;
        _options = options;
        _logger = logger;
    }

    public async Task<Job> CreateJobAsync(CreateJobRequest request, CancellationToken cancellationToken = default)
    {
        if (request is null)
        {
            throw new ValidationJobDispatchException("Request body is required.");
        }

        var type = GetRequiredType(request);
        var maxAttempts = ResolveMaxAttempts(request, _options.DefaultMaxAttempts);

        var job = new Job
        {
            Id = Guid.NewGuid(),
            Type = type,
            Name = string.IsNullOrWhiteSpace(request.Name) ? type : request.Name,
            Payload = ResolvePayload(request),
            MaxAttempts = maxAttempts,
            Status = JobStatus.Queued,
            CreatedAt = DateTimeOffset.UtcNow,
            UpdatedAt = DateTimeOffset.UtcNow,
            NextAttemptAt = DateTimeOffset.UtcNow
        };

        await _repository.AddAsync(job, cancellationToken);

        _logger.LogInformation("job_created job_id={JobId} type={Type} max_attempts={MaxAttempts}", job.Id, job.Type, job.MaxAttempts);
        return job;
    }

    public async Task<Job> GetJobAsync(Guid jobId, CancellationToken cancellationToken = default)
    {
        var job = await _repository.GetByIdAsync(jobId, cancellationToken);
        if (job is null)
        {
            throw new JobNotFoundException(jobId);
        }

        return job;
    }

    public async Task<Job> ClaimJobAsync(Guid jobId, ClaimJobRequest request, CancellationToken cancellationToken = default)
    {
        var job = await GetJobAsync(jobId, cancellationToken);
        var now = DateTimeOffset.UtcNow;
        var leaseOwner = string.IsNullOrWhiteSpace(request?.LeaseOwner) ? (string.IsNullOrWhiteSpace(request?.WorkerId) ? (string.IsNullOrWhiteSpace(request?.Worker) ? "worker" : request.Worker) : request.WorkerId) : request.LeaseOwner;
        var leaseDuration = request?.LeaseSeconds is null
            ? _options.LeaseDuration
            : TimeSpan.FromSeconds(request.LeaseSeconds.Value);

        if (leaseDuration <= TimeSpan.Zero)
        {
            throw new ValidationJobDispatchException("lease_seconds must be greater than zero.");
        }

        if (job.Status == JobStatus.Completed)
        {
            throw new InvalidJobStateException($"Job '{jobId}' is already completed.");
        }

        if (job.Status == JobStatus.Failed)
        {
            throw new InvalidJobStateException($"Job '{jobId}' is already failed.");
        }

        if (job.Status == JobStatus.Claimed && job.LeaseExpiresAt.HasValue && job.LeaseExpiresAt.Value > now)
        {
            throw new InvalidJobStateException($"Job '{jobId}' is already claimed until {job.LeaseExpiresAt:O}.");
        }

        if (job.Status == JobStatus.Queued || (job.Status == JobStatus.Claimed && (!job.LeaseExpiresAt.HasValue || job.LeaseExpiresAt.Value <= now)))
        {
            job.Status = JobStatus.Claimed;
            job.AttemptCount += 1;
            job.LeaseOwner = leaseOwner;
            job.LeaseExpiresAt = now.Add(leaseDuration);
            job.NextAttemptAt = null;
            job.UpdatedAt = now;
            await _repository.UpdateAsync(job, cancellationToken);

            _logger.LogInformation("job_claimed job_id={JobId} lease_owner={LeaseOwner} attempt_count={AttemptCount}", job.Id, job.LeaseOwner, job.AttemptCount);
            return job;
        }

        throw new InvalidJobStateException($"Job '{jobId}' cannot be claimed from status '{job.Status}'.");
    }

    public async Task<Job> CompleteJobAsync(Guid jobId, string? message = null, CancellationToken cancellationToken = default)
    {
        var job = await GetJobAsync(jobId, cancellationToken);
        var now = DateTimeOffset.UtcNow;

        if (job.Status == JobStatus.Completed)
        {
            return job;
        }

        if (job.Status != JobStatus.Claimed)
        {
            throw new InvalidJobStateException($"Job '{jobId}' must be claimed before it can be completed.");
        }

        job.Status = JobStatus.Completed;
        job.CompletedAt = now;
        job.UpdatedAt = now;
        job.LeaseOwner = null;
        job.LeaseExpiresAt = null;
        job.NextAttemptAt = null;
        job.LastError = message;

        await _repository.UpdateAsync(job, cancellationToken);

        _logger.LogInformation("job_completed job_id={JobId}", job.Id);
        return job;
    }

    public async Task<Job> FailJobAsync(Guid jobId, FailJobRequest request, CancellationToken cancellationToken = default)
    {
        var job = await GetJobAsync(jobId, cancellationToken);
        var now = DateTimeOffset.UtcNow;

        if (job.Status == JobStatus.Completed)
        {
            return job;
        }

        if (job.Status == JobStatus.Failed)
        {
            return job;
        }

        if (job.Status != JobStatus.Claimed)
        {
            throw new InvalidJobStateException($"Job '{jobId}' must be claimed before it can fail.");
        }

        var errorMessage = !string.IsNullOrWhiteSpace(request?.Error) ? request.Error : (!string.IsNullOrWhiteSpace(request?.Message) ? request.Message : "Job processing failed.");
        var transient = request is not null && (request.Transient ?? request.Retryable ?? false);
        var shouldRetry = transient && job.AttemptCount < job.MaxAttempts;

        if (shouldRetry)
        {
            job.Status = JobStatus.Queued;
            job.LeaseOwner = null;
            job.LeaseExpiresAt = null;
            job.LastError = errorMessage;
            job.NextAttemptAt = now.Add(_options.RetryBackoff);
            job.UpdatedAt = now;

            await _repository.UpdateAsync(job, cancellationToken);

            _logger.LogInformation("job_retry_scheduled job_id={JobId} retry_at={RetryAt}", job.Id, job.NextAttemptAt);
            return job;
        }

        job.Status = JobStatus.Failed;
        job.FailedAt = now;
        job.UpdatedAt = now;
        job.LeaseOwner = null;
        job.LeaseExpiresAt = null;
        job.NextAttemptAt = null;
        job.LastError = errorMessage;
        job.ErrorCode = "JOB_FAILED";

        await _repository.UpdateAsync(job, cancellationToken);

        _logger.LogInformation("job_failed job_id={JobId} error={Error}", job.Id, errorMessage);
        return job;
    }

    public async Task<IReadOnlyCollection<Job>> GetEligibleJobsAsync(CancellationToken cancellationToken = default)
    {
        return await _repository.GetEligibleJobsAsync(DateTimeOffset.UtcNow, cancellationToken);
    }

    public async Task<Job> ProcessQueuedJobAsync(Guid jobId, string? workerName = null, CancellationToken cancellationToken = default)
    {
        var job = await GetJobAsync(jobId, cancellationToken);

        if (job.Status == JobStatus.Completed || job.Status == JobStatus.Failed)
        {
            return job;
        }

        if (job.Status == JobStatus.Queued && job.NextAttemptAt.HasValue && job.NextAttemptAt.Value > DateTimeOffset.UtcNow)
        {
            return job;
        }

        var claimant = string.IsNullOrWhiteSpace(workerName) ? "background-worker" : workerName;
        var claimed = await ClaimJobAsync(jobId, new ClaimJobRequest { WorkerId = claimant, LeaseOwner = claimant, Worker = claimant }, cancellationToken);

        var shouldRetry = claimed.Type.Contains("fail", StringComparison.OrdinalIgnoreCase) || claimed.Name.Contains("fail", StringComparison.OrdinalIgnoreCase);
        if (shouldRetry)
        {
            return await FailJobAsync(jobId, new FailJobRequest { Error = "Worker failure simulated.", Transient = true }, cancellationToken);
        }

        if (claimed.Type.Contains("retry", StringComparison.OrdinalIgnoreCase) || claimed.Name.Contains("retry", StringComparison.OrdinalIgnoreCase))
        {
            return await FailJobAsync(jobId, new FailJobRequest { Error = "Transient retry required.", Transient = true }, cancellationToken);
        }

        return await CompleteJobAsync(jobId, "processed by worker", cancellationToken);
    }

    private static string GetRequiredType(CreateJobRequest request)
    {
        var type = request.Type ?? request.JobType ?? request.Name;
        if (string.IsNullOrWhiteSpace(type))
        {
            throw new ValidationJobDispatchException("Job type is required.");
        }

        return type.Trim();
    }

    private static object ResolvePayload(CreateJobRequest request)
    {
        if (request.Payload is not null)
        {
            return request.Payload;
        }

        if (request.Data is not null)
        {
            return request.Data;
        }

        if (request.Body is not null)
        {
            return request.Body;
        }

        return new { };
    }

    private static int ResolveMaxAttempts(CreateJobRequest request, int defaultMaxAttempts)
    {
        var candidate = request.MaxAttempts ?? request.Max_Attempts ?? defaultMaxAttempts;
        if (candidate <= 0)
        {
            throw new ValidationJobDispatchException("max_attempts must be greater than zero.");
        }

        return candidate;
    }
}

File: GoProject/internal/service/job_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/t-neubauer/port-csharp-to-go/GoProject/internal/domain"
	"github.com/t-neubauer/port-csharp-to-go/GoProject/internal/repository"
)

var ErrJobNotFound = errors.New("job not found")
var ErrValidation = errors.New("validation error")
var ErrInvalidState = errors.New("invalid job state")

func IsExpectedJobError(err error) bool {
	return errors.Is(err, ErrJobNotFound) ||
		errors.Is(err, ErrValidation) ||
		errors.Is(err, ErrInvalidState)
}

type Options struct {
	DefaultMaxAttempts int
	LeaseDuration      time.Duration
	RetryBackoff       time.Duration
}

type JobService struct {
	repository repository.JobRepository
	options    Options
}

func NewJobService(repo repository.JobRepository, options Options) *JobService {
	return &JobService{repository: repo, options: options}
}

func (s *JobService) CreateJob(ctx context.Context, request domain.CreateJobRequest) (domain.Job, error) {
	jobType := strings.TrimSpace(request.Type)
	if jobType == "" {
		jobType = strings.TrimSpace(request.JobType)
	}
	if jobType == "" {
		jobType = strings.TrimSpace(request.Name)
	}
	if jobType == "" {
		return domain.Job{}, fmt.Errorf("%w: job type is required", ErrValidation)
	}

	maxAttempts := s.options.DefaultMaxAttempts
	if request.MaxAttempts != nil {
		maxAttempts = *request.MaxAttempts
	}
	if maxAttempts <= 0 {
		return domain.Job{}, fmt.Errorf("%w: max_attempts must be greater than zero", ErrValidation)
	}

	payload := request.Payload
	if len(payload) == 0 {
		payload = request.Data
	}
	if len(payload) == 0 {
		payload = request.Body
	}
	if len(payload) == 0 {
		payload = json.RawMessage(`{}`)
	}

	now := time.Now().UTC()
	id, err := domain.NewID()
	if err != nil {
		return domain.Job{}, err
	}
	job := domain.Job{
		ID:            id,
		JobType:       jobType,
		Name:          request.Name,
		Payload:       payload,
		Status:        domain.StatusQueued,
		MaxAttempts:   maxAttempts,
		CreatedAt:     now,
		UpdatedAt:     now,
		NextAttemptAt: &now,
	}
	if job.Name == "" {
		job.Name = jobType
	}
	return s.repository.Add(ctx, job)
}

func (s *JobService) GetJob(ctx context.Context, id string) (domain.Job, error) {
	job, ok, err := s.repository.GetByID(ctx, id)
	if err != nil {
		return domain.Job{}, err
	}
	if !ok {
		return domain.Job{}, fmt.Errorf("%w: %s", ErrJobNotFound, id)
	}
	return job, nil
}

func (s *JobService) ClaimJob(ctx context.Context, id string, request domain.ClaimJobRequest) (domain.Job, error) {
	job, err := s.GetJob(ctx, id)
	if err != nil {
		return domain.Job{}, err
	}
	now := time.Now().UTC()
	if job.Status == domain.StatusCompleted || job.Status == domain.StatusFailed {
		return domain.Job{}, fmt.Errorf("%w: job is already terminal", ErrInvalidState)
	}
	if job.Status == domain.StatusClaimed && job.LeaseExpiresAt != nil && job.LeaseExpiresAt.After(now) {
		return domain.Job{}, fmt.Errorf("%w: job is already claimed", ErrInvalidState)
	}
	if job.Status == domain.StatusQueued && job.NextAttemptAt != nil && job.NextAttemptAt.After(now) {
		return domain.Job{}, fmt.Errorf("%w: retry is not due", ErrInvalidState)
	}
	leaseDuration := s.options.LeaseDuration
	if request.LeaseSeconds != nil {
		leaseDuration = time.Duration(*request.LeaseSeconds) * time.Second
	}
	if leaseDuration <= 0 {
		return domain.Job{}, fmt.Errorf("%w: lease_seconds must be greater than zero", ErrValidation)
	}
	owner := strings.TrimSpace(request.LeaseOwner)
	if owner == "" {
		owner = strings.TrimSpace(request.WorkerID)
	}
	if owner == "" {
		owner = strings.TrimSpace(request.Worker)
	}
	if owner == "" {
		owner = "worker"
	}
	expires := now.Add(leaseDuration)
	job.Status = domain.StatusClaimed
	job.AttemptCount++
	job.LeaseOwner = owner
	job.LeaseExpiresAt = &expires
	job.NextAttemptAt = nil
	job.UpdatedAt = now
	return s.repository.Update(ctx, job)
}

func (s *JobService) CompleteJob(ctx context.Context, id string, request domain.CompleteJobRequest) (domain.Job, error) {
	job, err := s.GetJob(ctx, id)
	if err != nil {
		return domain.Job{}, err
	}
	if job.Status == domain.StatusCompleted {
		return job, nil
	}
	if job.Status != domain.StatusClaimed {
		return domain.Job{}, fmt.Errorf("%w: job must be claimed before completion", ErrInvalidState)
	}
	now := time.Now().UTC()
	job.Status = domain.StatusCompleted
	job.CompletedAt = &now
	job.UpdatedAt = now
	job.LeaseOwner = ""
	job.LeaseExpiresAt = nil
	job.NextAttemptAt = nil
	job.LastError = request.Message
	return s.repository.Update(ctx, job)
}

func (s *JobService) FailJob(ctx context.Context, id string, request domain.FailJobRequest) (domain.Job, error) {
	job, err := s.GetJob(ctx, id)
	if err != nil {
		return domain.Job{}, err
	}
	if job.Status == domain.StatusCompleted || job.Status == domain.StatusFailed {
		return job, nil
	}
	if job.Status != domain.StatusClaimed {
		return domain.Job{}, fmt.Errorf("%w: job must be claimed before failure", ErrInvalidState)
	}
	now := time.Now().UTC()
	transient := request.Transient != nil && *request.Transient
	if request.Retryable != nil && *request.Retryable {
		transient = true
	}
	if transient && job.AttemptCount < job.MaxAttempts {
		next := now.Add(s.options.RetryBackoff)
		job.Status = domain.StatusQueued
		job.NextAttemptAt = &next
		job.LeaseOwner = ""
		job.LeaseExpiresAt = nil
		job.LastError = failureMessage(request)
		job.UpdatedAt = now
		return s.repository.Update(ctx, job)
	}
	job.Status = domain.StatusFailed
	job.FailedAt = &now
	job.NextAttemptAt = nil
	job.LeaseOwner = ""
	job.LeaseExpiresAt = nil
	job.LastError = failureMessage(request)
	job.ErrorCode = "JOB_FAILED"
	job.UpdatedAt = now
	return s.repository.Update(ctx, job)
}

func (s *JobService) GetEligibleJobs(ctx context.Context) ([]domain.Job, error) {
	return s.repository.GetEligible(ctx, time.Now().UTC())
}

func (s *JobService) ProcessQueuedJob(ctx context.Context, id string, workerName string) (domain.Job, error) {
	job, err := s.GetJob(ctx, id)
	if err != nil {
		return domain.Job{}, err
	}
	if job.Status == domain.StatusCompleted || job.Status == domain.StatusFailed {
		return job, nil
	}
	if job.Status == domain.StatusQueued && job.NextAttemptAt != nil && job.NextAttemptAt.After(time.Now().UTC()) {
		return job, nil
	}

	claimed, err := s.ClaimJob(ctx, id, domain.ClaimJobRequest{
		LeaseOwner: workerName,
	})
	if err != nil {
		return domain.Job{}, err
	}

	switch {
	case strings.Contains(claimed.JobType, "fail") || strings.Contains(claimed.Name, "fail"):
		transient := true
		return s.FailJob(ctx, id, domain.FailJobRequest{
			Error:     "worker failure simulated.",
			Transient: &transient,
		})
	case strings.Contains(claimed.JobType, "retry") || strings.Contains(claimed.Name, "retry"):
		transient := true
		return s.FailJob(ctx, id, domain.FailJobRequest{
			Error:     "transient retry required.",
			Transient: &transient,
		})
	default:
		return s.CompleteJob(ctx, id, domain.CompleteJobRequest{
			Message: "processed by worker",
		})
	}
}

func failureMessage(request domain.FailJobRequest) string {
	if strings.TrimSpace(request.Error) != "" {
		return request.Error
	}
	if strings.TrimSpace(request.Message) != "" {
		return request.Message
	}
	return "job processing failed"
}
